File: src/java/billiards/codeseq/InvalidCodeSequence.java
package billiards.codeseq;

import org.eclipse.collections.api.list.primitive.IntList;

public enum InvalidCodeSequence {
    EMPTY,
    NEGATIVE_OR_ZERO_NUMBERS,
    ILLEGAL_PATTERN;

    public static String errorMessage(
        final IntList codeNumbers, final InvalidCodeSequence errorCode) {
        final String str = codeNumbers.makeString(" ");

        final String msg;
        if (errorCode == InvalidCodeSequence.EMPTY) {
            msg = String.format("Code sequence is empty: %s", str);
        } else if (errorCode == InvalidCodeSequence.NEGATIVE_OR_ZERO_NUMBERS) {
            msg = String.format("All code numbers must be greater than 0: %s", str);
        } else if (errorCode == InvalidCodeSequence.ILLEGAL_PATTERN) {
            msg = String.format("Code sequence is illegal: %s", str);
        } else {
            throw new RuntimeException("Unknown InvalidCodeSequence " + errorCode);
        }

        return msg;
    }
}

File: src/backend/cpp/invalid_code_sequence.cpp
#include "invalid_code_sequence.hpp"



/*
Jul 31 2025 Marco Mai
transfer from Java
*/
inline std::string makeString(const std::vector<int>& codeNumbers, const std::string& sep ) {
    std::ostringstream oss;
    for (size_t i = 0; i < codeNumbers.size(); ++i) {
        if (i > 0) oss << sep;
        oss << codeNumbers[i];
    }
    return oss.str();
}

inline std::string errorMessage(const std::vector<int>& codeNumbers, InvalidCodeSequence errorCode) {
    std::string str = makeString(codeNumbers);

    switch (errorCode) {
        case InvalidCodeSequence::EMPTY:
            return "Code sequence is empty: " + str;
        case InvalidCodeSequence::NEGATIVE_OR_ZERO_NUMBERS:
            return "All code numbers must be greater than 0: " + str;
        case InvalidCodeSequence::ILLEGAL_PATTERN:
            return "Code sequence is illegal: " + str;
        default:
            throw std::runtime_error("Unknown InvalidCodeSequence");
    }
}
